libraries/chain/include/eos/chain/authority_checker.hpp

```diff
@@ -0,0 +1,127 @@
+ *
+ * To determine whether an authority is satisfied or not, we first determine which keys have approved of a message, and
+ * then determine whether that list of keys is sufficient to satisfy the authority. This class takes a list of keys and
+ * provides the @ref satisfied method to determine whether that list of keys satisfies a provided authority.
+ *
+ * @tparam F A callable which takes a single argument of type @ref AccountPermission and returns the corresponding
+ * authority
+ */
+template<typename F>
+class AuthorityChecker {
+   F PermissionToAuthority;
+   vector<public_key_type> signingKeys;
+   vector<bool> usedKeys;
+
+   struct WeightTallyVisitor {
+      using result_type = UInt32;
+
+      AuthorityChecker& checker;
+      UInt32 totalWeight = 0;
+
+      WeightTallyVisitor(AuthorityChecker& checker)
+         : checker(checker) {}
+
+      UInt32 operator()(const types::KeyPermissionWeight& permission) {
+         auto itr = boost::find(checker.signingKeys, permission.key);
+         if (itr != checker.signingKeys.end()) {
+            checker.usedKeys[itr - checker.signingKeys.begin()] = true;
+            totalWeight += permission.weight;
+         }
+         return totalWeight;
+      }
+      UInt32 operator()(const types::AccountPermissionWeight& permission) {
+         //TODO: Recursion limit? Yes: implement as producer-configurable parameter
+         if (checker.satisfied(permission.permission))
+            totalWeight += permission.weight;
+         return totalWeight;
+      }
+   };
+
+public:
+   AuthorityChecker(F PermissionToAuthority, const flat_set<public_key_type>& signingKeys)
+      : PermissionToAuthority(PermissionToAuthority),
+        signingKeys(signingKeys.begin(), signingKeys.end()),
+        usedKeys(signingKeys.size(), false)
+   {}
+
+   bool satisfied(const types::AccountPermission& permission) {
+      return satisfied(PermissionToAuthority(permission));
+   }
+   template<typename AuthorityType>
+   bool satisfied(const AuthorityType& authority) {
+      // Save the current used keys; if we do not satisfy this authority, the newly used keys aren't actually used
+      auto KeyReverter = fc::make_scoped_exit([this, keys = usedKeys] () mutable {
+         usedKeys = keys;
+      });
+
+      // Sort key permissions and account permissions together into a single set of MetaPermissions
+      detail::MetaPermissionSet permissions;
+      permissions.insert(authority.keys.begin(), authority.keys.end());
+      permissions.insert(authority.accounts.begin(), authority.accounts.end());
+
+      // Check all permissions, from highest weight to lowest, seeing if signingKeys satisfies them or not
+      WeightTallyVisitor visitor(*this);
+      for (const auto& permission : permissions)
+         // If we've got enough weight, to satisfy the authority, return!
+         if (permission.visit(visitor) >= authority.threshold) {
+            KeyReverter.cancel();
+            return true;
```

Comment:
I think we can make the over-authorization a "soft" consensus rather than a "hard" consensus.  In other words, block producers may choose to include over-authorized transactions, but in practice will reject them. This will allow us to update / refine this without hard forking the chain.